src/utils/database-common/src/transactions/db_transaction_manager.rs

```diff
@@ -34,49 +34,74 @@ pub trait DatabaseTransactionManager: Send + Sync {
 
 /////////////////////////////////////////////////////////////////////////////////////////
 
-pub async fn run_transactional<H, HFut>(
-    base_catalog: &Catalog,
-    callback: H,
-) -> Result<(), InternalError>
-where
-    H: FnOnce(Catalog) -> HFut + Send + Sync + 'static,
-    HFut: std::future::Future<Output = Result<(), InternalError>> + Send + 'static,
-{
-    // Extract transaction manager, specific for the database
-    let db_transaction_manager = base_catalog
-        .get_one::<dyn DatabaseTransactionManager>()
-        .unwrap();
-
-    // Start transaction
-    let transaction_ref = db_transaction_manager.make_transaction_ref().await?;
-
-    // Create a chained catalog for transaction-aware components,
-    // but keep a local copy of a transaction pointer
-    let chained_catalog = CatalogBuilder::new_chained(base_catalog)
-        .add_value(transaction_ref.clone())
-        .build();
-
-    // Run transactional code in the callback
-    let result = callback(chained_catalog).await;
-
-    // Commit or rollback transaction depending on the result
-    match result {
-        // In case everything succeeded, commit the transaction
-        Ok(_) => {
-            db_transaction_manager
-                .commit_transaction(transaction_ref)
-                .await?;
-            Ok(())
-        }
+pub struct DatabaseTransactionRunner<HFutResultE = InternalError> {
+    _phantom: PhantomData<HFutResultE>,
+}
 
-        // Otherwise, do an explicit rollback
-        Err(e) => {
-            db_transaction_manager
-                .rollback_transaction(transaction_ref)
-                .await?;
-            Err(e)
+impl<HFutResultE> DatabaseTransactionRunner<HFutResultE> {
```

Comment:
Can we make `DatabaseTransactionRunner` a dill component? This way a top-level service could inject it instead of a `catalog` and do:
```rust
self.tx_runner.transactional(|tx_catalog| { ... });
```
It's a small usability improvement, but may simplify testing top-level services.

As a step further in usability we could hide catalog entirely as such:
```rust
self.tx_runner.transactional(|my_service: Arc<dyn MyService>| my_service.do_stuff())
```
i.e. by closure argument type we can have runner extract the right component from the catalog and pass it into the transactional context.

Reply:
very interesting proposition :thinking: